src/examples/libpmem2/ringbuf/ringbuf.c

```diff
@@ -0,0 +1,433 @@
+/* Copyright 2019-2021, Intel Corporation */
+
+/*
+ * ringbuf.c -- a simple lock-free single producer single consumer ring buffer
+ *	implemented using libpmem2.
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <libpmem2.h>
+
+#define RINGBUF_SIGNATURE "my_fast_ringbuf"
+#define RINGBUF_SIGNATURE_LEN (sizeof(RINGBUF_SIGNATURE))
+
+#define RINGBUF_POS_PERSIST_BIT (1ULL << 63)
+
+/*
+ * Persistent on-media format of the ring buffer.
+ *
+ * Fields are carefully aligned and padded. This is done to avoid: a) flushing
+ * metadata fields when unnecessary, and b) misaligned non-temporal writes.
+ */
+struct ringbuf_data {
+	uint8_t signature[RINGBUF_SIGNATURE_LEN]; /* 0 - 16 */
+
+	uint64_t nentries; /* 16 - 24 */
+
+	uint64_t entry_size; /* 24 - 32 */
+
+	uint8_t padding[32]; /* 32 - 64 */
+
+	struct {
+		uint64_t read;	/* 64 - 72 */
+		uint64_t write; /* 72 - 80 */
+	} pos;
+	uint8_t padding2[48]; /* 80 - 128 */
+
+	uint8_t data[]; /* 128 - */
+};
+
+/* on-media format of a single entry. nothing but data. */
+struct ringbuf_entry {
+	uint8_t data[0];
+};
+
+/* runtime (ephemeral) ring buffer state */
+struct ringbuf {
+	struct pmem2_map *map;
+
+	pmem2_persist_fn persist;
+	pmem2_memcpy_fn memcpy;
+	enum pmem2_granularity granularity;
+
+	struct ringbuf_data *data;
+};
+
+/*
+ * ringbuf_data_is_initialized -- checks whether the ring buffer data has
+ * initialized signature in the header. If so, it's assumed that the
+ * rest of the on-media format is valid as well.
+ */
+static int
+ringbuf_data_is_initialized(struct ringbuf_data *rbuf_data)
+{
+	return memcmp(rbuf_data->signature, RINGBUF_SIGNATURE,
+			RINGBUF_SIGNATURE_LEN) == 0;
+}
+
+/*
+ * ringbuf_data_force_page_allocation -- ensures that all the pages for
+ * the ring buffer data are allocated by rewriting a byte from each page.
+ * This serves two purposes: a) eliminating the kernel page allocation overheads
+ * from the time measurements, and b) making sure that the process won't be
+ * killed due to lack of space.
+ */
+static void
+ringbuf_data_force_page_allocation(struct ringbuf_data *rbuf_data, size_t size)
+{
+	long pagesize = sysconf(_SC_PAGESIZE);
+	volatile char *cur_addr = (char *)rbuf_data;
+	char *addr_end = (char *)cur_addr + size;
+	for (; cur_addr < addr_end; cur_addr += pagesize) {
+		*cur_addr = *cur_addr;
+	}
+}
+
+/*
+ * ringbuf_initialize_data -- initializes the header of the on-media
+ * format in a fail-safe atomic manner.
+ * The data is written out and persisted first, followed by the signature.
+ */
+static void
+ringbuf_initialize_data(struct ringbuf *rbuf, uint64_t nentries,
+			uint64_t entry_size)
+{
+	rbuf->data->pos.write = 0;
+	rbuf->data->pos.read = 0;
+	rbuf->data->nentries = nentries;
+	rbuf->data->entry_size = entry_size + sizeof(struct ringbuf_entry);
+	rbuf->persist(rbuf->data, sizeof(*rbuf->data));
+
+	rbuf->memcpy(rbuf->data->signature, RINGBUF_SIGNATURE,
```

Comment:
What about the signature? Shouldn't be persisted here? If not, ringbuf will be initialized each run.

Reply:
this is a pmem memcpy that includes a flush.